File: interviewbit-heaps-and-maps-distinct-numbers-in-window.py
class Solution:
    # @param A : list of integers
    # @param B : integer
    # @return a list of integers
    def dNums(self, A, B):
        if B > len(A):
            return []

        window = {}
        for num in A[:B]:
            window[num] = window.get(num, 0) + 1

        distinct = [len(window)]
        for i in range(len(A) - B):
            remkey = A[i]
            if window[remkey] == 1:
                del window[remkey]
            else:
                window[remkey] -= 1

            addkey = A[i + B]
            window[addkey] = window.get(addkey, 0) + 1
            distinct.append(len(window))

        return distinct

File: interviewbit-heaps-and-maps-distinct-numbers-in-window.java
public class Solution {
    public int[] dNums(int[] A, int B) {
        if (B > A.length) {
            return new int[0];
        }
        int[] distinct = new int[A.length - B + 1];
        HashMap<Integer, Integer> counts = new HashMap<>();
        for (int i = 0; i < B; i++) {
            counts.put(A[i], counts.getOrDefault(A[i], 0) + 1);
        }
        distinct[0] = counts.size();
        for (int i = 0; i < A.length - B; i++) {
            if (counts.get(A[i]) == 1) {
                counts.remove(A[i]);
            } else {
                counts.put(A[i], counts.get(A[i]) - 1);
            }
            counts.put(A[i + B], counts.getOrDefault(A[i + B], 0) + 1);
            distinct[i + 1] = counts.size();
        }
        return distinct;
    }
}
